### PR TITLE
Bugfix/issue 1828 Good text failing

### LINKED ISSUE #1828: Setting bad data in one T&G field then good data quickly in another can lead to the good data failing.
### Bug Summary

This bug is related to iOS Issue #[1781](https://github.com/smartdevicelink/sdl_ios/issues/1781)

Based on how the Text & Graphic manager currently works, it is possible to have good data fail. A possible solution is that when a `Show `fails and the `currentScreenData `gets updated to also update the `updateState` of the T&G operation as well. This will require knowing which fields the T&G operation is updating so that they don't also get reset.

With a similar method we could potentially also stop sending all Show data and only send updated Show data.

### Reproduction Steps

1. Set "bad" data (data that will get rejected) into `TextField1` and start the operation.
2. Set "good" data (data that will not get rejected) into `TextField2` and start the operation before the first operation gets a Show response.

Expected Behavior: `TextField1 `should remain in its original state and `TextField2` should show the new data.

Observed Behavior: `TextField1` and `TextField2` will show the original state data because `TextField2`'s operation includes the `TextField1` data as well.

### Which projects have you seen this bug on?

_No response_

### Android Version(s)

_No response_

### Android Device(s)

n/a

### sdl_java_suite Version

5.5.0

### Testing Environment(s)

```markdown
Manticore
```

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by ChloeMJM**
Fixes #1828

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).
- [x] I have tested Android, *Java SE, and *Java EE

#### Unit Tests
*Note: This branch has been tested for Android by Chloe, Julian Tested JavaSE

- `testOnShowFailBadDataDoesNotUpdateScreen()` : Tests the show failure branch within the `sendShow` method which notifies the `currentScreenDataUpdateListener` of an `errorState`.
- `testUpdateTargetStateWithErrorStateNullDoesNotUpdateCurrentScreen()`: Tests the `updateTargetStateWithErrorState` method to ensure that fields set to null will not be updated on the current screen.
- `testUpdateTargetStateWithErrorBadDataDoesNotUpdateCurrentScreen()` : Tests the `updateTargetStateWithErrorState` method to ensure that fields set to bad data will not be updated on the current screen.
- `testUpdateTargetStateWithErrorBadDataAndGoodData()` : Tests the `updateTargetStateWithErrorState` method to ensure that fields set to bad data will not be updated on the current screen but fields set with good data will be updated.

#### Core Tests
In Manticore verified that properties in screen manager when set would not fail when one attribute was given bad data

### Summary
- In `TextAndGraphicUpdateOperation:` modified the `sendShow` method to call `currentScreenDataUpdateListener.onError(updatedState)` on failure.
- Modified the `onError` path of the `currentScreenDataUpdateListener` override to call `updatePendingOperationsWithFailedScreenState(errorState)`, passing in an `errorState`.
- In `BaseTextAndGraphicManager`: added in method `updatePendingOperationsWithFailedScreenState `which sends the `errorState` to `updateOperation.updateTargetStateWithErrorState(errorState)`
- In `TextAndGraphicUpdateOperation`: added method `updateTargetStateWithErrorState` that ensures every field is null safe and compares the `errorState` to the `updatedState`. If they are the same, then it sets the `updatedState` to the `currentScreenData `associated with the field to keep the field from being updated with bad data.
- Added unit tests.

##### Enhancements
* Screen data will update even if other fields may have bad data.

##### Bug Fixes
* Screen data will update even if other fields may have bad data.

### Tasks Remaining
- [x] Create fix.
- [x] Add unit tests.
- [x] Code review.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Review by JulianKast — CHANGES_REQUESTED** [review]
First round code review

File: base/src/main/java/com/smartdevicelink/managers/screen/BaseTextAndGraphicManager.java [hunk 1]
```diff
@@ -217,9 +194,10 @@ public void onUpdate(TextAndGraphicState newScreenData) {
             }
 
             @Override
-            public void onError() {
+            public void onError(TextAndGraphicState errorState) {
                 // Invalidate data that's different from our current screen data
                 resetFieldsToCurrentScreenData();
+                updatePendingOperationsWithFailedScreenState(errorState);
```
[the remark this answers is not in the document]
Comment (ChloeMJM) [reply base/src/main/java/com/smartdevicelink/managers/screen/BaseTextAndGraphicManager.java]: Okay, I applied this for now, let me know if they say anything different. 
Also applied all other feedback in commit #8.

**Review by JulianKast — CHANGES_REQUESTED** [review]
Don't stress about code coverage, given the design, it would require a lot of rewriting of current unit tests to get coverage in areas we did not have before to get to 60%

File: base/src/main/java/com/smartdevicelink/managers/screen/TextAndGraphicUpdateOperation.java [hunk 2]
```diff
@@ -154,11 +202,10 @@ public void onResponse(int correlationId, RPCResponse response) {
             internalInterface.get().sendRPC(show);
         } else {
             DebugTool.logInfo(TAG, "ISdl is null Text and Graphic update failed");
-            currentScreenDataUpdateListener.onError();
+            currentScreenDataUpdateListener.onError(null);
```
[the remark this answers is not in the document]
Comment (ChloeMJM) [reply base/src/main/java/com/smartdevicelink/managers/screen/TextAndGraphicUpdateOperation.java]: Suggestion applied

File: base/src/main/java/com/smartdevicelink/managers/screen/TextAndGraphicUpdateOperation.java [hunk 3]
```diff
@@ -171,7 +218,7 @@ public void onResponse(int correlationId, RPCResponse response) {
                     updateCurrentScreenDataFromSetDisplayLayout(setLayout);
                 } else {
                     DebugTool.logInfo(TAG, "Text and Graphic SetDisplayLayout failed");
-                    currentScreenDataUpdateListener.onError();
+                    currentScreenDataUpdateListener.onError(null);
```
[the remark this answers is not in the document]
Comment (ChloeMJM) [reply base/src/main/java/com/smartdevicelink/managers/screen/TextAndGraphicUpdateOperation.java]: Suggestion applied

File: base/src/main/java/com/smartdevicelink/managers/screen/TextAndGraphicUpdateOperation.java [hunk 4]
```diff
@@ -180,7 +227,7 @@ public void onResponse(int correlationId, RPCResponse response) {
             internalInterface.get().sendRPC(setLayout);
         } else {
             DebugTool.logInfo(TAG, "ISdl is null Text and Graphic update failed");
-            currentScreenDataUpdateListener.onError();
+            currentScreenDataUpdateListener.onError(null);
```
[the remark this answers is not in the document]
Comment (ChloeMJM) [reply base/src/main/java/com/smartdevicelink/managers/screen/TextAndGraphicUpdateOperation.java]: Suggestion applied

**Review by JulianKast — APPROVED** [review]

**Pull request merged**